src/Rank.cpp

```diff
@@ -797,6 +804,17 @@ void Rank::addToPipes(
 
 		// if there are any matching attacks we add them
 		if (!pipeAttacksToAdd.IsEmpty()) {
+
```

Comment:
I think that all this (in more or less every place it's done) and the relevant macros perhaps would be better done as a public function of a Pipe. Then one could just ask the Pipe if it has conflicting isTremulant values in the attacks and save a few repeated lines. The function would just return a bool value which in turn would decide whether to trigger the warning message or not.

Reply:
And possibly fire that message dialog itself as well...

Reply:
> And possibly fire that message dialog itself as well...

That had occurred to me too.  I'm not a C++ programmer.  Even though I certainly understand and agree with what you're saying (public function of Pipe, etc.) implementation might take me some figuring out.

Reply:
Actually, as a message dialog would block the execution, it could be better for the flow to just log the warning and proceed.